client/src/www/model/platform_error.ts

```diff
@@ -0,0 +1,156 @@
+ */
+function convertRawErrorObjectToPlatformError(rawObj: object): PlatformError {
+  if (!rawObj) {
+    return newInvalidJSONPlatformError('error JSON is falsy', rawObj);
+  }
+
+  if (!('code' in rawObj) || typeof rawObj.code !== 'string') {
+    return newInvalidJSONPlatformError('`code` is not string in error JSON', rawObj);
+  }
+  const code = rawObj.code.trim();
+  if (!code) {
+    return newInvalidJSONPlatformError('`code` is empty in error JSON', rawObj);
+  }
+
+  if (!('message' in rawObj) || typeof rawObj.message !== 'string') {
+    return newInvalidJSONPlatformError('`message` is not string in error JSON', rawObj);
+  }
+
+  let options: {details?: object; cause?: Error} = null;
+  if ('details' in rawObj || 'cause' in rawObj) {
+    options = {};
+    if ('details' in rawObj && typeof rawObj.details === 'object' && rawObj.details) {
+      options.details = rawObj.details;
+    }
+    if ('cause' in rawObj && typeof rawObj.cause === 'object' && rawObj.cause) {
+      options.cause = convertRawErrorObjectToPlatformError(rawObj.cause);
+    }
+  }
+
+  return new PlatformError(code, rawObj.message, options);
+}
+
+/**
+ * PlatformError is used to communicate error details from Go to TypeScript.
+ */
+export class PlatformError extends CustomError {
+  public readonly details: object = null;
+
+  /**
+   * Constructs a new PlatformError instance with the specified parameters.
+   * @param {ErrorCode} code An ErrorCode representing the category of this error.
+   * @param {string} message A user-readable string of this error.
+   * @param options An object containing the optional details and cause.
+   */
+  constructor(
+    public readonly code: ErrorCode,
+    message: string,
+    options?: {
+      details?: object;
+      cause?: Error;
+    }
+  ) {
+    super(message, options);
+    this.details = options?.details;
+  }
+
+  /**
+   * Parses a JSON string into a {@link PlatformError}.
+   *
+   * If {@link rawJSON} is invalid, it returns a PlatformError of {@link InvalidLogic} with the
+   * original JSON in its details.
+   * @param {string} rawJSON The JSON string to parse.
+   * @returns {PlatformError} A non-null PlatformError object.
+   */
+  static parseJSON(rawJSON: string): PlatformError {
+    let rawObj: unknown;
+    try {
+      rawObj = JSON.parse(rawJSON);
+    } catch {
+      return newInvalidJSONPlatformError('invalid error JSON string', rawJSON);
+    }
+    if (typeof rawObj !== 'object') {
+      return newInvalidJSONPlatformError('error JSON is not an object', rawJSON);
+    }
+    return convertRawErrorObjectToPlatformError(rawObj);
+  }
+
+  /**
+   * Returns a user readable string of this error with all details and causes.
+   * @returns {string} A user friendly string representing this error.
+   */
+  public toString(): string {
+    let result = this.code + '\n' + this.message;
+    if (this.details) {
+      result += '\nDetails: ';
```

Comment:
Not for the error details, that's more for the service providers or us instead of the users.

And if we do, we need to translate the error strings in Go as well.